man/lxpanelxctl.1
'\" t
.\"     Title: LXPANELX
.\"    Author: Ying-Chun Liu
.\" Generator: DocBook XSL Stylesheets v1.75.2 <http://docbook.sf.net/>
.\"      Date: March  2, 2008
.\"    Manual: http://LXDE.org
.\"    Source: http://LXDE.org
.\"  Language: English
.\"
.TH "LXPANELX" "1" "March 2, 2008" "http://LXDE\&.org" "http://LXDE.org"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
lxpanelxctl \- controller for lxpanel\&.
.SH "SYNOPSIS"
.HP \w'\fBlxpanelctl\fR\ 'u
\fBlxpanelctl\fR {command}
.SH "DESCRIPTION"
.PP
This manual page documents briefly the
\fBlxpanelctl\fR
command\&.
.PP
\fBlxpanelctl\fR
is a program that controls lxpanel\&.
.SH "COMMANDS"
.PP
\fBmenu\fR
.RS 4
Show the system menu\&.
.RE
.PP
\fBrun\fR
.RS 4
Show the run dialog\&.
.RE
.PP
\fBconfig\fR
.RS 4
Show the panel configuration dialog\&. If more than one panel is configured, it is undefined which panel configuration dialog will be shown\&.
.RE
.PP
\fBrestart\fR
.RS 4
Restart lxpanel\&.
.RE
.PP
\fBexit\fR
.RS 4
Exit lxpanel\&.
.RE
.SH "SEE ALSO"
.PP
lxpanel (1)\&.
.SH "AUTHOR"
.PP
This manual page was written by paulliu
grandpaul@gmail\&.com
for the
Debian
system (but may be used by others)\&. Permission is granted to copy, distribute and/or modify this document under the terms of the
GNU
General Public License, Version 2 any later version published by the Free Software Foundation\&.
.PP
On Debian systems, the complete text of the GNU General Public License can be found in /usr/share/common\-licenses/GPL\&.
.SH "AUTHOR"
.PP
\fBYing\-Chun Liu\fR
.RS 4
Author.
.RE
.SH "COPYRIGHT"
.br
Copyright \(co 2008 paulliu
.br


debian/changelog
lxpanelx (0.6.0+svn20120717-1) unstable; urgency=low

  * Initial 0.6 build

 -- Vadim Ushakov <igeekless@gmail.com>  Tue, 17 Jul 2012 09:28:35 +0000

lxpanelx (0.5.6+svn20111104-1) unstable; urgency=low

  * Initial lxpanelx deb build

 -- Vadim Ushakov <igeekless@gmail.com>  Fri, 04 Nov 2011 22:48:43 +0800

lxpanel (0.5.5-1) unstable; urgency=low

  * New upstream release
    - Startup becomes slightly faster. (might not be visible to users)
    - Configuration dialog of launchbar applet become much faster. (quite
      prominent)
    - Translations update
    - Several minor bug fixes 
    - Fixed crash problem with libmenu-cache >= 0.3.1
  * Updated upstream url to http://lxde.org
  * Bumped Standard-version to 3.8.4

 -- Andrew Lee (李健秋) <ajqlee@debian.org>  Fri, 19 Feb 2010 15:37:59 +0100

lxpanel (0.5.4.1-1) unstable; urgency=low

  * New upstream release (Closes: #493253, #534222, #548333, #560843)
    - Fixed crash problem on creating new panel
    - Fixed segfault when battery is not exist 
    - Use _NET_SHOWING_DESKTOP to toggle the desktop efficiently in wincmd 
      applet when available
    - Better checks for EWMH compliant window managers 
    - Fixed cpufreq plugin, thermal plugin (Closes: #535568)
    - Support recognizing current DE with $XDG_CURRENT_DESKTOP
    - Fixed string format warning in src/misc.c
    - Translation updated
  * debian/control:
    - drop deprecated lxpanel-netstat-plugin binary package
  * Added README.source
  * Updated README.Debian
  * Bumped Standard-version to 3.8.3

 -- Andrew Lee (李健秋) <ajqlee@debian.org>  Sat, 19 Dec 2009 00:34:40 +0800

lxpanel (0.5.3-1) experimental; urgency=low

  * New upstream release
    - Added documentation of \n to dclock dialog
    - Added configuration dialog to deskno plugin
    - Improved usability on narrow vertical panel support, now it's possible
      to configure a vertical panel as small as 16 pixels.
  * debian/control:
    - Recommends on manpages-dev to be able to have 'man 3 strftime' to
      check the format codes of digital clock plugin. (Closes:#493245)

 -- Andrew Lee (李健秋) <ajqlee@debian.org>  Mon, 10 Aug 2009 03:13:50 +0800

lxpanel (0.5.2-1) experimental; urgency=low

  * New upstream release
    - Fixed "Run" dialog hangs problem.
    - Restored the old behavior of taskbar in earlier versions
    - Hided signal strength display in the icon of the netstatus applet 
      if the monitored iface is not wireless.
    - Improved application launcher, Available apps are now categorized 
      as they are in applications menu.
    - The class name of the desktop panels are now changed to 
      "PanelToplevel". This is compatible with gnome and can be used by
      theme authors to distinguish the panel widgets from others.
    - Reduced program size.
  * debian/control:
    - Build-depends on xsltproc

 -- Andrew Lee (李健秋) <ajqlee@debian.org>  Fri, 07 Aug 2009 22:20:31 +0800

lxpanel (0.5.0-1) experimental; urgency=low

  * New upstream release
    - Change behavior of preference dialogs
    - Added window icons to configuration dialogs
    - Added help on dclock format strings
    - Added keyboard equivalents to taskbar popup menu (FR2781949)
    - Fixed width control range, fix geometry calculations so it
      doesn't fly off edge of screen
    - Improved geometry of menu configuration dialog
    - Replaced xterm by lxterminal as default terminal
    - Switch desktop on taskbar button Click and Raise (Bug2028953)
    - Keyboard equivalents on taskbar popup submenu (FR2781949)
    - Repaired bad function signature on focus-out callback
    - Rewrited WM_STRUT setting so it doesn't update the property 
      unless it changed
    - Fixed the panel transparency colorbox
    - Fixed CPU plugin loses context on a relayout (Patch 2800828, song6song)
    - Added autohide
    - Rewrote battery backend code (based on Linux acpi 1.4)
    - Added scroll wheel on main icon to adjust volume and middle-click 
      on main icon to adjust mute
    - Use Docbook-XML instead of SGML
    - Added --enable-man for manpage regeneration
    - Added New wincmd icon from http://www.oxygen-icons.org/
    - Icon size now configurable
    - Enable xkb plugin
    - Added logic to have a visible name on grouped tasks
    - Added way to get line breaks in dclock formatting
    - Rework system tray to do event handling at the X protocol level
    - Rework background draw to solve observed systray corruption
    - Rework 90 degree re-orientation to solve lost tray icons for
      Qt applications, conditional on 2.16 or later
    - Fix memory leaks
  * debian/rules:
    - Added --enable-man
  * debian/control:
    - Build-depends on docbook-xsl

 -- Andrew Lee (李健秋) <ajqlee@debian.org>  Wed, 29 Jul 2009 20:56:08 +0200

lxpanel (0.4.1+svn20090524-2) unstable; urgency=low

  * Update debian/control (Thanks Jonas Smedegaard's for the advice):
    - Move conflicts lxnm ( <0.3 ) from lxpanel-netstat-plugin to lxpanel
    - Make lxpanel-netstat-plugin dummy package depends on connman
      (See ITP #475022 for detail)
  * Removed deprecated dh_desktop from debian/rules
  * Bumped Standard-version to 3.8.2

 -- Andrew Lee <andrew@linux.org.tw>  Sun, 19 Jul 2009 13:48:27 +0200

lxpanel (0.4.1+svn20090524-1) unstable; urgency=low

  * New upstream release (Closes: #516603, #511001, #493242, #486284, 
    #515633, #484606, #495924, #517858)
    - A whole new application menu generated according to 
      freedesktop.org menu spec. (faster and more standard-compliant. 
      has tooltips for menu items)
    - Enhanced "Run" dialog.
    - Auto resize of application launcher buttons when panel size gets 
      changed.
    - The problematic netstat plugin is now turned off by default. It 
      will be moved to a separate project later.
    - Building dynamic panel plugins outside the source tree of lxpanel 
      is now possible.
    - Fix icon display problems under 64-bit systems.

src/lxpanelxctl.c
/**
 * Copyright (c) 2006 LxDE Developers, see the file AUTHORS for details.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

/* Remote controller of lxpanelx */

#include "lxpanelxctl.h"
#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <X11/Xatom.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static Display* dpy;

static const char usage[] =
        "\nlxpanelxctl - LXPanelX Controller\n"
        "Usage: lxpanelxctl <command>\n\n"
        "Available commands:\n"
        "menu\tshow system menu\n"
        "run\tshow run dialog\n"
        "config\tshow configuration dialog\n"
        "restart\trestart lxpanelx\n"
        "exit\texit lxpanelx\n\n";
/*
static int get_cmd( const char* cmd )
{
    if( ! strcmp( cmd, "menu") )
        return LXPANELX_CMD_SYS_MENU;
    else if( ! strcmp( cmd, "run") )
        return LXPANELX_CMD_RUN;
    else if( ! strcmp( cmd, "config") )
        return LXPANELX_CMD_CONFIG;
    else if( ! strcmp( cmd, "restart") )
        return LXPANELX_CMD_RESTART;
    else if( ! strcmp( cmd, "exit") )
        return LXPANELX_CMD_EXIT;
    return -1;
}*/

int main( int argc, char** argv )
{
    char *display_name = (char *)getenv("DISPLAY");
    XEvent ev;
    Window root;
    Atom cmd_atom;
//    int cmd = -1;
    /* int restart; */

    if( argc < 2 )
    {
        printf( usage );
        return 1;
    }
/*
    if (argc == 2)
        cmd = get_cmd(argv[1]);

    if (cmd != -1)
    {
        dpy = XOpenDisplay(display_name);
        if (dpy == NULL) {
            printf("Cant connect to display: %s\n", display_name);
            exit(1);
        }
        root = DefaultRootWindow(dpy);
        cmd_atom = XInternAtom(dpy, "_LXPANELX_CMD", False);
        memset(&ev, '\0', sizeof ev);
        ev.xclient.type = ClientMessage;
        ev.xclient.window = root;
        ev.xclient.message_type = cmd_atom;
        ev.xclient.format = 8;

        ev.xclient.data.b[0] = cmd;

        XSendEvent(dpy, root, False,
                   SubstructureRedirectMask|SubstructureNotifyMask, &ev);
        XSync(dpy, False);
        XCloseDisplay(dpy);
    }
    else*/
    {
        size_t buff_size = 0;
        int i;
        for (i = 1; i < argc; i++)
        {
            buff_size += strlen(argv[i]) + 1;
        }

        char * buff = (char *) calloc(buff_size, sizeof(char));

        size_t buff_pos = 0;
        for (i = 1; i < argc; i++)
        {
            size_t s = strlen(argv[i]) + 1;
            memcpy(buff + buff_pos, argv[i], s * sizeof(char));
            buff_pos += s;
        }

        
        dpy = XOpenDisplay(display_name);
        if (dpy == NULL) {
            printf("Cant connect to display: %s\n", display_name);
            exit(1);
        }
        root = DefaultRootWindow(dpy);
        cmd_atom = XInternAtom(dpy, "_LXPANELX_TEXT_CMD", False);

        Atom type_atom = XInternAtom(dpy, "UTF8_STRING", False);

        XChangeProperty (dpy, root, cmd_atom, type_atom, 8, PropModeReplace,
                         (unsigned char *) buff, buff_size);
        XSync(dpy, False);
        XCloseDisplay(dpy);
    }

    return 0;
}

